### PR TITLE
Use new inner constructor and type declaration syntax

### PR DESCRIPTION
**Pull request by andreasnoack**
The new syntax for inner constructors doesn't parse on 0.5 so it will be ugly and more demanding to support on both versions. Furthermore, 0.6 will be out soon and nobody will care about 0.5. If they care anyway and we add features then we can backport those features to a 0.5 branch.

### CODE REVIEW

**Comment by andyferris** [comment]
LGTM.  Thanks Andreas. 

Just so you know, I was planning on merging and releasing #105 first, before merging this one. This plus major changes due to #106 are required before this package is usable on v0.6. Those other changes are backward compatible so maybe they should be released for v0.5 too (otherwise it would block backporting). 

Would a `@static if VERSION < ...` version of this PR be a sensible approach? Or is it better all-round to focus development on a single version of Julia?

**Comment by andyferris** [comment]
CI is complaint about `TupleN`

**Comment by andreasnoack** [comment]
The problem is that the syntax doesn't parse at all so we'll end up with a lot of definitions like https://github.com/Evizero/UnicodePlots.jl/blob/master/src/graphics/bargraphics.jl#L1-L44 which I think is ugly. Personally, I think it is better to focus development hours on 0.6 and thereafter instead of compatibility.

There is a with the build bots so CI is running with old binaries that don't have the latest syntax changes. Tests pass locally on latest master.

**Comment by c42f** [comment]
I think merging this might be a bit premature - are the change here required to make the package work on 0.6, or are they syntactic niceties?  If possible it would be nice for master to continue working with 0.5 for a while.

**Comment by andreasnoack** [comment]
The inner constructor syntax is deprecated so it gives warnings on master. The only way to make it work across 0.5 and 0.6 is to write that code as strings and use `include_string` conditionally on the version such as in the link I posted above.

**Comment by timholy** [comment]
There is always the `(::Type{MyNewType{T}}){T}(arg) = new{T}(arg)` syntax, which works on both.

**Comment by andreasnoack** [comment]
Thanks for pointing that out. I wasn't aware of that. I'll update the PR.

**Comment by andyferris** [comment]
Thanks Andreas (and good idea, Tim).

I still think it would be better to delay this for just the moment and first address #106, which completely and utterly destroys *any* custom subtypes of `StaticArray`, is a much larger amount of work, and the fix is compatible with Julia v0.5.

At that point we can base development on v0.6 (and backport crucial fixes to v0.5 - but not fixing #106 first will make any backports less trivial). This will let us (in no particular order):

 * Not require `Compat`.
 * Make use of `where` syntax to simplify dispatch (e.g. [this](https://github.com/JuliaArrays/StaticArrays.jl/pull/105#discussion_r101935226)).
 * Let us take advantage of compiler (specifically, inference and constant propagation) improvements (this will mean, less `@generated` functions).
 * Keep code cleaner.

Basically, this package is unusually low-level and reliant on certain compiler behavior - I'd rather not have this package be artificially limited by continuing to support two versions of Julia.

I do apologize about the deprecation warning noise between now and then - I would suggest to keep rebasing this branch on the other changes for the next little while, and use this branch for downstream development work targeting v0.6 in the meantime.

I am also open to dropping v0.5 support *now* if people think this is a better route (how many backports will there really be?). But please, do keep in mind how completely broken custom static arrays are - deprecation warnings might be a *good* warning-sign for users :)

**Comment by andreasnoack** [comment]
I don't really understand why merging these small deprecation warning fixes could be a problem for fixing custom static arrays. Notice, that the PR doesn't drop 0.5 anymore.

**Comment by andyferris** [comment]
I suppose it should be OK now that I'm tagging the `similar_type` changes. I'll be ripping out `Compat` sometime soon when we go to v0.6-only.

@andreasnoack this PR now has conflicts (my fault)

**Comment by c42f** [comment]
Nice, thanks for putting up with 0.5 for a little longer :-)

File: test/fixed_size_arrays.jl [hunk 1]
```diff
@@ -8,15 +8,15 @@ import StaticArrays.FixedSizeArrays: @fixed_vector
...
-typealias Vec1d Vec{1, Float64}
-typealias Vec2d Vec{2, Float64}
-typealias Vec3d Vec{3, Float64}
-typealias Vec4d Vec{4, Float64}
-typealias Vec3f Vec{3, Float32}
-
-typealias Mat2d Mat{2,2, Float64, 4}
-typealias Mat3d Mat{3,3, Float64, 9}
-typealias Mat4d Mat{4,4, Float64, 16}
+@compat const Vec1d = Vec{1, Float64}
```
Comment (andyferris) [inline test/fixed_size_arrays.jl]: The sometimes `const` and sometimes not syntax isn't really great (not your fault, mind you).

Comment (c42f) [inline test/fixed_size_arrays.jl]: Ah yes, I noticed this syntax quirk too.  Hard to know what to do about it.  In packages, it would be nice if all bindings at global scope were implicitly const by default.  But for simple scripting that would be fairly horrible.

Comment (andyferris) [inline test/fixed_size_arrays.jl]: Could be a special property of `Main`? (Or some property that could be set on a module-by-module basis).

**Comment by andyferris** [comment]
Travis errors are *odd*...

File: src/SMatrix.jl [hunk 2]
```diff
@@ -66,7 +66,7 @@ end
         SMatrix{S1, S2, $T, L}(x)
     end
 end
-typealias SMatrixNoType{S1, S2, L, T} SMatrix{S1, S2, T, L}
+@compat SMatrixNoType{S1, S2, L, T} = SMatrix{S1, S2, T, L}
```
Comment (andyferris) [inline src/SMatrix.jl]: I don't know why, but I don't remember this existing, which is why I couldn't fathom the test failure... In theory, `MMatrix` should be symmetric. But don't worry about this for this PR.

**Comment by andreasnoack** [comment]
On 0.6, there seems to be an issue with `@inferred` and dot broadcast syntax. Looking into it now.

**Comment by andyferris** [comment]
I made an issue for that

**Comment by andyferris** [comment]
https://github.com/JuliaLang/julia/issues/20620

**Comment by andreasnoack** [comment]
Oh, I missed that one. I think there is an easy fix.

**Comment by andyferris** [comment]
I just merged this, Andreas. I still plan to remove *Compat* and v0.5 support sometime soonish.

**Comment by c42f** [comment]
Yes, let's make a release before we break everything with #113 (I guess this will end up dropping 0.5 support... honestly I was hoping to last a little longer!)

**Comment by andyferris** [comment]
I was wondering who the target audience was? There is no functional changes for v0.5 users and v0.6 users will be using quite a different codebase soon... @andreasnoack is using `master` for v0.6 development reasonable for the time-being?

**Comment by andreasnoack** [comment]
Using a release is slightly simpler in a larger group of collaborators but the main reason is Travis testing where it is pretty inconvenient to use master.

**Comment by andyferris** [comment]
Yes, Travis. That makes sense.
